mona_openai/mona_openai.py

```diff
@@ -0,0 +1,243 @@
+    #   isn't inited yet once we have the relevant endpoint for this.
+
+    class MonitoredOpenAI(base_class):
+        """
+        A mona-monitored version of an openai API class.
+        """
+
+        @classmethod
+        def _get_mona_single_message(
+            cls, kwargs_param, start_time, is_exception, is_async, response
+        ):
+            """
+            Returns a MonaSingleMessage object to be used for data
+            exporting to Mona's servers by a Mona client.
+            """
+            # Recreate the input dict to avoid manipulating the caller's data,
+            # and remove Mona-related data.
+            new_kwargs = deepcopy(
+                {
+                    x: kwargs_param[x]
+                    for x in kwargs_param
+                    if not x.startswith(MONA_ARGS_PREFIX)
+                }
+            )
+
+            message = {
+                "input": new_kwargs,
+                "latency": time.time() - start_time,
+                "is_exception": is_exception,
+                "api_name": COMPLETION_CLASS_NAME,
+                "is_async": is_async,
+            }
+
+            if ADDITIONAL_DATA_ARG_NAME in kwargs_param:
+                message["additional_data"] = kwargs_param[
+                    ADDITIONAL_DATA_ARG_NAME
+                ]
+
+            if response:
+                message["response"] = response
+                message["analysis"] = super()._get_analysis_params(
+                    new_kwargs, response, specs
+                )
+
+            message = super()._get_clean_message(message, specs)
+
+            return MonaSingleMessage(
+                message=message,
+                contextClass=context_class,
+                contextId=kwargs_param.get(
+                    CONTEXT_ID_ARG_NAME, response["id"] if response else None
+                ),
+                exportTimestamp=kwargs_param.get(
+                    EXPORT_TIMESTAMP_ARG_NAME, start_time
+                ),
+            )
+
+        @classmethod
+        async def _inner_create(
+            cls, export_function, super_function, args, kwargs
+        ):
+            """
+            The main logic for wrapping create functions with mona data
+            exporting.
+            This internal function porovides a template for both sync
+            and async activations (helps with wrapping both "create"
+            and "acreate").
+            """
+            start_time = time.time()
+
+            response = None
+
+            async def _inner_mona_export(is_exception):
+                await call_non_blocking_sync_or_async(
+                    export_function,
+                    (
+                        cls._get_mona_single_message(
+                            kwargs,
+                            start_time,
+                            is_exception,
+                            super_function.__name__ == "acreate",
```

Comment:
It depends if we got here from "create" or "acreate", no?

Reply:
Oh man this just confused me as well. Note that this is a equality check and not a param assignment. What you're describing is exactly what we're testing here...